### DISCUSSION TITLE
FPCA

### DISCUSSION THREAD

**Discussion by aroarmath** [GAA-UAM/scikit-fda#362] [Q&A] [answered]
Hi! I am exploring the library to work with functional principal components. I have seen how to calculate them, but I can't find the way to express the data based on two of the components for example. Could you help me?

**Comment by vnmabus** [comment, marked as answer]
Hello there! The `transform` method of the `FPCA` class gives the coefficients of each curve (with the mean subtracted) in the basis of the principal components.

You can recover a smoothed version of the original curve expressing the linear combination manually (we should probably add a method for this in the future):

```python
import skfda

X, y = skfda.datasets.fetch_weather(return_X_y=True)
X = X.coordinates[0]

fpca = skfda.preprocessing.dim_reduction.feature_extraction.FPCA(n_components=5)
transf = fpca.fit_transform(X)

fig = X[0].plot()
(fpca.mean_ + (fpca.components_ * transf[0]).sum()).plot(fig=fig).show()
```
![Figure_1](https://user-images.githubusercontent.com/2364173/127227558-0af7f0b1-2fe7-4899-a1d3-ca729fb1668a.png)

**Reply by aroarmath** [reply]
Thanks, but i am searching for a representation where one of the axis would be principal component 1 and the otre axis the principal component two.

**Reply by aroarmath** [reply]
Here you have and example in R language with biplot https://rpubs.com/Joaquin_AR/287787

**Reply by vnmabus** [reply]
Well, once you `transform` your data you have a vector of `n_components` features. You can use the `scatter` function in Matplotlib to represent the two first coefficients in a scatterplot, or use another visualization method for multivariate data. See https://stackoverflow.com/questions/39216897/plot-pca-loadings-and-loading-in-biplot-in-sklearn-like-rs-autoplot for an example in scikit-learn which should be similar.

**Reply by aroarmath** [reply]
And if we have functional data, could we have this kind of representation? Because de n_componentes obtained from a fpca method has not the same type of a multivariate method.

**Reply by aroarmath** [reply]
Sorry, yesterday I uploaded a multivariate example, but this one is for functional data
https://www.google.com/url?sa=t&source=web&rct=j&url=https://dialnet.unirioja.es/descarga/articulo/6765740.pdf&ved=2ahUKEwixrtXiiIXyAhW86uAKHW_dDMkQFjAKegQIBxAC&usg=AOvVaw2QDUrdtHpwZuFV7RBQ6Eqx

**Reply by vnmabus** [reply]
You mean the "Análisis de correspondencias" graph? Isn't it a scatterplot of the coefficients of the first two principal components?
An example of the thing I was proposing is the following:
```python
import skfda
import matplotlib.pyplot as plt

X, y = skfda.datasets.fetch_weather(return_X_y=True)
X = X.coordinates[0]

fpca = skfda.preprocessing.dim_reduction.feature_extraction.FPCA(n_components=2)
transf = fpca.fit_transform(X)

plt.scatter(transf[:, 0], transf[:, 1])
for i, txt in enumerate(X.sample_names):
    plt.annotate(txt, (transf[i, 0], transf[i, 1]))
plt.show()
```
![Figure_1](https://user-images.githubusercontent.com/2364173/127294401-f48d6f2e-2537-4d5f-bcd8-fe8809b52278.png)

**Reply by aroarmath** [reply]
It was really helpfull, thanks for all coments 😊😊😊
